lec09-modeling-r.Rmd

```diff
@@ -1,4 +1,615 @@
+state <- c(X = 2, Y = 2)
+times <- seq(0, 50, by = 0.01)
+
+result <- ode(y = state, times = times, func = predator_prey, parms = parameters)
+result <- data.frame(result)
+
+```
+
+Plot a trajectory in the phase plane and the null clines:
+
+```{r}
+library(ggplot2)
+
+x_seq <- seq(0, 8, by = 0.1)
+y_seq <- seq(0, 8, by = 0.1)
+
+df <- data.frame(x_seq, y_seq)
+
+p1 <- ggplot() + # assign plot to a variable so we can use it later
+  geom_path(aes(x = X, y = Y, color = time), result) + # trajectory
+  geom_line(aes(x = 0, y = y_seq), df, color = 'dark red') + # x null cline 1
+  geom_line(aes(x = x_seq, y = a/b), df, color = 'dark red') # x null cline 2
+
+p1
+```
+
+The null clines go through the places where the trajectory turns back on itself -- they go through
+the 'peaks' and 'valleys' of $x$. 
+
+Another way to think of the null clines is as if we were back to a 1D system and were 
+plotting $dx/dt$ vs. $x$ for a series of values of $y$. The null cline passes through all
+the points where $dx/dt = 0$. 
+
+Now let's find the null clines for $y$ by solving $0 = cx - dy$. In general, we want an equation in the
+form $y(x)$, $y$ as a function of $x$, so that we can more easily plot it on our phase portrait axes. 
+
+After rearraning, we find the solution to this equation: $y = \frac{c}{d}x$. Let's add it to the plot
+
+```{r}
+p1 +
+  geom_line(aes(x = x_seq, y = x_seq*c/d), df, colour = 'blue') +
+  xlim(0, 8) +
+  ylim(0, 8)
+```
+
+The intersection(s) of the $x$ and $y$ null clines are the fixed points, so plotting null
+clines is also a way to graphically find the fixed points. This is particularly useful if 
+it's hard to solve for the fixed points analytically. 
+
+The `phaseR` package also includes a `nullclines` function to draw nullclines for you.
+
+```{r, eval=FALSE}
+?nullclines
+```
+
+```{r}
+# plot vector field: on a grid of points, plot an arrow in the direction of dx/dt and dy/dt
+pp_flowField <- flowField(predator_prey, x.lim = c(0,10), y.lim = c(0,10),
+                          parameters = params,
+                          points = 15, # this is the density of grid points on which to plot arrows
+                          system = 'two.dim', # 'two.dim' is default
+                          add = FALSE)
+
+# add trajectories
+pp_trajectory <- trajectory(predator_prey, 
+                            # y0 is a matrix where each row is pairs of (X, Y) 
+                            y0 = initial_conditions, 
+                            t.end = 30, # how far in time to calculate the trajectories
+                            parameters = params, system = "two.dim")
+
+pp_nullclines <- nullclines(predator_prey, 
+           x.lim = c(0, 10), # axis limits
+           y.lim = c(0, 10),
+           # y0 is a matrix where each row is pairs of (X, Y) 
+           y0 = initial_conditions, 
+           t.end = 30, # how far in time to calculate the trajectories 
+           parameters = params, system = "two.dim")
+```
```

Comment:
Why is the red line not covering the entire range of `y` when `x=0`? And why is there that little dip in when `x=1` and `y=5`?

Reply:
I also noticed that dip, I don't think it's real though because if you change the scale it doesn't change size. I don't know why it's there, or why the red line doesn't go all the way to y = 0.